src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,33 +124,143 @@ def __init__(
+        if self._messages_ready_since_last_commit >= self._max_batch_size:
+            return True
+        if self._last_commit_time + self._max_batch_time <= datetime.now():
+            return True
+        return False
+
+    def _bulk_commit(self) -> None:
+        """Commits and clears the ready to commit queue."""
+        if self._messages_ready_since_last_commit == 0:
+            return
+        self._commit(self._ready_to_commit)
+        self._clear_ready_queue()
+
+    def _clear_ready_queue(self) -> int:
+        """Clears the ready to commit queue and returns the amount of items dropped."""
+        amount_dropped = self._messages_ready_since_last_commit
+        self._ready_to_commit = {}
+        self._messages_ready_since_last_commit = 0
+        self._last_commit_time = datetime.now()
+        return amount_dropped
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        deadline = time.time() + timeout if timeout else None
+        self._bulk_commit()
+        while self._ongoing_billing_outcomes:
+            if deadline and deadline > time.time():
+                items_left = self._messages_ready_since_last_commit + len(
+                    self._ongoing_billing_outcomes
+                )
+                logger.warning(f"join timed out, items in the queue left: {items_left}")
+                break
+            self._mark_commit_ready()
+            self._bulk_commit()
+
     def terminate(self) -> None:
         self.close()
+        amount_dropped = len(self._ongoing_billing_outcomes)
+        while self._ongoing_billing_outcomes:
+            ongoing_work, _ = self._ongoing_billing_outcomes.popleft()
+            ongoing_work.cancel()
+        amount_dropped += self._clear_ready_queue()
+        if amount_dropped > 0:
+            logger.warning(f"terminated, items dropped: {amount_dropped}")
 
     def close(self) -> None:
-        self.__closed = True
+        self._closed = True
 
     def submit(self, message: Message[KafkaPayload]) -> None:
-        assert not self.__closed
-        self.__messages_since_last_commit += 1
+        if self._closed:
+            raise MessageRejected("Consumer is closed, no new messages are accepted.")
 
-        payload = self._get_payload(message)
-        self._produce_billing_outcomes(payload)
-        self._mark_commit_ready(message)
+        bucket_payload = self._get_bucket_payload(message)
+        quantity = self._count_processed_transactions(bucket_payload)
 
-    def _get_payload(self, message: Message[KafkaPayload]) -> MetricsBucket:
+        if quantity < 1:
+            # We still want to commmit buckets that don't generate billing
+            # outcomes, since the consumer has already processed them. To keep
+            # the offset order when committing, we still need to go through the
+            # futures queue.
+            future: Future[Message[KafkaPayload]] = Future()
+            future.set_result(True)  # Mark the future as completed
```

Comment:
Initializing a `Future` has some overhead (it initializes a condition variable which in turn initializes a lock and a deque as far as I can see), so I would prefer if we could just set `future = None` here (and handle that case accordingly in `_mark_commit_ready`).

https://github.com/python/cpython/blob/1863302d61a7a5dd8b8d345a00f0ee242c7c10bf/Lib/concurrent/futures/_base.py#L330

Reply:
It seems I updated the code while you were reviewing :D

TIL about these internals of `Future`.